### ISSUE TITLE
pgwire/hba: HBA rule for "local" can't be applied to new node when joining an existing cluster

### ISSUE THREAD

**Issue by a-robinson** [cockroachdb/cockroach#49652]
**Describe the problem**

New nodes joining a cluster that have a HBA policy set for "local" connections will fail to apply the policy the first time that they start up and join the cluster. Attempts to connect via the new node's unix socket will require a password even if the cluster-wide HBA policy is set to trust such connections.

**To Reproduce**

1. Create a cluster using v20.1.0 (1 node is fine).
2. Run `set cluster setting server.host_based_authentication.configuration = 'local all all trust';` to set up an HBA policy that trusts all connections made via unix socket.
3. Start up an additional node and have it join the cluster.

At this point attempting to connect to the latter node's unix socket will fail unless you provide a password. Although you don't even need to specify a unix socket directory to see the failure -- just look in the logs for the new node and you'll see:

```
W200528 16:21:56.923714 316 server/settingsworker.go:100  [n2] setting "server.host_based_authentication.configuration" to "local all all trust" failed: candidate pg code: 55000
  - error with attached stack trace:
    github.com/cockroachdb/cockroach/pkg/sql/pgwire.checkHBASyntaxBeforeUpdatingSetting
        /go/src/github.com/cockroachdb/cockroach/pkg/sql/pgwire/hba_conf.go:148
    github.com/cockroachdb/cockroach/pkg/settings.(*StringSetting).Validate
        /go/src/github.com/cockroachdb/cockroach/pkg/settings/string.go:57
    github.com/cockroachdb/cockroach/pkg/settings.(*StringSetting).set
        /go/src/github.com/cockroachdb/cockroach/pkg/settings/string.go:65
    github.com/cockroachdb/cockroach/pkg/settings.updater.Set
        /go/src/github.com/cockroachdb/cockroach/pkg/settings/updater.go:92
    github.com/cockroachdb/cockroach/pkg/server.(*Server).refreshSettings.func1
        /go/src/github.com/cockroachdb/cockroach/pkg/server/settingsworker.go:99
    github.com/cockroachdb/cockroach/pkg/server.(*Server).refreshSettings.func2
        /go/src/github.com/cockroachdb/cockroach/pkg/server/settingsworker.go:116
    github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunWorker.func1
        /go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:198
    runtime.goexit
        /usr/local/go/src/runtime/asm_amd64.s:1357
  - error with embedded safe details: authentication rule type 'local' requires all nodes to be upgraded to %s
    -- arg 1: <roachpb.Version>
  - authentication rule type 'local' requires all nodes to be upgraded to 19.2-8
```

**Expected behavior**

I expect the HBA cluster setting to be applied to new nodes that join the cluster.

**Environment:**

 - CockroachDB version: v20.1.0

cc @knz

**Comment by knz** [comment]
nice catch!

@tbg the issue here is that the logic must wait until the version setting has propagated before it adopts the config change for another setting. Do we have a "barrier sync" operator to wait for all settings to propagate?

**Comment by tbg** [comment]
No, not currently, though @ajwerner had a similar problem the other day. It seemed reasonably easy to fix, we can delay serving the pgL listener until we've seen a settings update in Gossip.

**Comment by a-robinson** [comment]
I assume that this and the related issues aren't likely to get any attention for 20.2?

**Comment by knz** [comment]
The issue #48273 still needs attention unfortunately.

**Comment by a-robinson** [comment]
Hey guys, any estimate on if/when #48273 and this may get prioritized? It's pretty frustrating to have newly initialized nodes (or nodes that are coming back up after having their disk wiped) not functional until they're either restarted or have a manual `SET CLUSTER SETTINGS` command run.

Although maybe you can refresh my memory -- would a node running v20.2 be able to use this setting that was added in v20.1 without needing to know the versions of all the other nodes in the cluster? If this particular instance of the problem will stop popping up once all nodes are running v20.2, I'd be much less concerned about the timing of a fix.

**Comment by a-robinson** [comment]
If the release candidate is anything to go by, it looks like this particular setting won't cause problems on v20.2 👍

**Comment by tbg** [comment]
@a-robinson sorry this has been causing trouble. There's a PR out that makes sure nodes don't lose their cluster setting across restarts (https://github.com/cockroachdb/cockroach/pull/55166) though that wouldn't have helped you here, as you're looking specifically at the case of a brand new node. But, since we have the Join RPC now, something similar can be done at joining without any kind of artificial waits. I have suggested that as a follow-up on #55166 and will use this issue to track it.
In the meantime, if you see this kind of problem again, the easiest remedy is probably just a 5s sleep - not sure if you'd be opposed to that; it sure is annoying to have to do that.

**Comment by a-robinson** [comment]
No worries, Tobi. I'm not sure I understand your 5s sleep suggestion, though. Sleep before doing what?

To make sure it's clear, the problem is that if this error is hit during the startup of a new node, then that new node will not apply the proper settings until it's either restarted or a `SET CLUSTER SETTING` command is run. If neither of those manual actions are taken, then the intended value of the `server.host_based_authentication.configuration` setting is never actually applied on the new node, even after hours or days.

**Comment by tbg** [comment]
Oh, I didn't understand the issue until now. Thanks for the clarification. At this point, is it fair to say that the issue is resolved as of 20.2? I agree with you that this particular migration will not matter any more at that point, so the specific issue you've seen should no longer pop up.

Additionally, the work I mentioned around the join RPC should avoid similar problems in the future.

**Comment by a-robinson** [comment]
Sorry, I didn't make that very clear in the original report.

Yeah, we can close this in favor of the other related issues. I don't 100% trust my local tests, but I'm able to repro the problem on 20.1.8 and not on 20.2.0-rc.4, so it's more likely than not that the problem isn't there on v20.2 (since the newly started v20.2 node can trust that all nodes in the cluster are at least at v20.1).
